src/gretel_trainer/relational/core.py

```diff
@@ -180,17 +316,104 @@ def remove_foreign_key(self, table: str, constrained_columns: List[str]) -> None
         self._clear_safe_ancestral_seed_columns(table)
 
     def update_table_data(self, table: str, data: pd.DataFrame) -> None:
-        try:
-            self.graph.nodes[table]["metadata"].data = data
-            self.graph.nodes[table]["metadata"].columns = set(data.columns)
-            self._clear_safe_ancestral_seed_columns(table)
-        except KeyError:
-            raise MultiTableException(
-                f"Unrecognized table name: {table}. If this is a new table to add, use `add_table`."
-            )
+        if table in self.relational_jsons:
+            _, original_pk, original_fks = self._remove_relational_json(table)
+            new_rj_ingest = RelationalJson.ingest(table, original_pk, data)
+            if new_rj_ingest is not None:
+                self._add_rel_json_and_tables(table, new_rj_ingest)
+                parent_table_name = new_rj_ingest[0].root_table_name
+            else:
+                self._add_single_table(
+                    name=table,
+                    primary_key=original_pk,
+                    data=data,
+                )
+                parent_table_name = table
+            for fk in original_fks:
+                self.add_foreign_key(
+                    table=fk.table_name,
+                    constrained_columns=fk.columns,
+                    referred_table=parent_table_name,
+                    referred_columns=fk.parent_columns,
+                )
+        else:
+            try:
+                metadata = self.graph.nodes[table]["metadata"]
+                new_rj_ingest = RelationalJson.ingest(table, metadata.primary_key, data)
+                if new_rj_ingest is not None:
+                    original_foreign_keys = self._get_user_defined_fks_to_table(table)
+                    self.graph.remove_node(table)
+                    self._add_rel_json_and_tables(table, new_rj_ingest)
+                    for fk in original_foreign_keys:
+                        self.add_foreign_key(
+                            table=fk.table_name,
+                            constrained_columns=fk.columns,
+                            referred_table=new_rj_ingest[0].root_table_name,
+                            referred_columns=fk.parent_columns,
+                        )
+                else:
+                    metadata.data = data
+                    metadata.columns = set(data.columns)
+                    self._clear_safe_ancestral_seed_columns(table)
+            except KeyError:
+                raise MultiTableException(
+                    f"Unrecognized table name: {table}. If this is a new table to add, use `add_table`."
+                )
```

Comment:
nit: IMO this `try:except` could be a bit more local, with this much code in between, it's hard to understand where this error can be raised.

Reply:
Good call, fixed